### PR TITLE
File functions edit 2.1

### LINKED ISSUE #72: save_data_to_file() in test_memory_backend fails on Windows
https://github.com/oasis-open/cti-taxii-server/blob/b9441e14b73eefbd5c3b85c6de7a1c4e4db81a8a/medallion/test/test_memory_backend.py#L396-L403

> save_data_to_file(f.name) attempts to open the temporary file, which has already been opened by NamedTemporaryFile.  That doesn't work (at least on Windows), and causes a permission error.
>
> Python docs say: "Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows NT or later)."
>
> https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile
>
> I think save_data_to_file() should probably be changed to accept a file-like object rather than a filename, akin to json.dump().

Found by @chisholm.

### PR DESCRIPTION
**Pull request by zrush-mitre**
Added the ability for save and load functions to take either strings or file pointers, and changed the test for saving files to resolve #72

### CODE REVIEW

File: medallion/test/test_memory_backend.py [hunk 1]
```diff
@@ -536,15 +536,19 @@ def test_saving_data_file(self):  # just for the memory backend
...
                 data=json.dumps(new_bundle),
                 headers=post_header,
             )
-            self.app.medallion_backend.save_data_to_file(f.name)
+            self.app.medallion_backend.save_data_to_file(f)
             assert os.path.isfile(f.name)
 
+            # seek(0) is needed to reset the file pointer so it's correctly
+            # read by the backend again
+            f.seek(0)
```
Comment (chisholm) [inline medallion/test/test_memory_backend.py]: The file was opened for writing; you mustn't try to read from it.  I think you probably want mode="w+".

**Comment by chisholm** [comment]
In the memory backend load/save file methods, there is no way of setting an encoding when a filename is given.  Do we need that?  One advantage of just supporting a textual file-like object is that you don't have to deal with that.  The stream was opened in textual mode so someone has made that decision already.  All you have to do is read/write from/to it.

Also, regarding the `type(filename) is str` code: should we support `unicode` too, for the sake of Python 2?

**Comment by clenk** [comment]
Yes, generally we should use 

```python
from six import string_types

if isinstance(filename, string_types):
```

instead of

```python
if type(filename) is str:
```

I think it's fine to support both filename and file-like object - if people need to set encoding themselves they can use the file-like object. But maybe we should default to `utf-8` when just a filename is provided?

**Comment by chisholm** [comment]
Yeah, definitely prefer `isinstance`.  True, we could let encoding default when a filename is used, since we have that second way to do it.  Okay, utf-8 sounds better than Windows' typical default of cp1252 anyway.

File: medallion/backends/memory_backend.py [hunk 2]
```diff
@@ -104,13 +107,19 @@ def _pop_expired_sessions(self):
             self.next.pop(item)
 
     def load_data_from_file(self, filename):
-        with open(filename, "r") as infile:
-            self.data = json.load(infile)
+        if isinstance(filename, string_types):
+            with io.open(filename, "r", encoding="utf-8") as infile:
```
Comment (emmanvg) [inline medallion/backends/memory_backend.py]: could we place the encoding in the function itself and pass it to io.open(). In the event someone wishes to use a different one?

File: medallion/backends/memory_backend.py [hunk 3]
```diff
@@ -104,13 +107,19 @@ def _pop_expired_sessions(self):
...
+                self.data = json.load(infile)
+        else:
+            self.data = json.load(filename)
 
     def save_data_to_file(self, filename, **kwargs):
         """The kwargs are passed to ``json.dump()`` if provided."""
-        with open(filename, "w") as outfile:
-            json.dump(self.data, outfile, **kwargs)
+        if isinstance(filename, string_types):
+            with io.open(filename, "w", encoding="utf-8") as outfile:
```
Comment (emmanvg) [inline medallion/backends/memory_backend.py]: could we place the encoding in the function itself and pass it to io.open(). In the event someone wishes to use a different one?

File: medallion/backends/memory_backend.py (hunk 2 above)
Comment (chisholm) [reply medallion/backends/memory_backend.py]: You mean like

```python
encoding="utf-8"
io.open(filename, "r", encoding=encoding)
```
where the encoding is in a variable?  I don't see what that accomplishes?  Maybe I don't understand the suggestion.

Comment (emmanvg) [reply medallion/backends/memory_backend.py]: No, I meant more like:

``` python
def load_data_from_file(self, filename, encoding="utf-8"):
    if isinstance(filename, string_types):
        with io.open(filename, "r", encoding=encoding) as infile:
```

Comment (chisholm) [reply medallion/backends/memory_backend.py]: That was in my original [comment](https://github.com/oasis-open/cti-taxii-server/pull/99#issuecomment-587852116).  @clenk subsequently [decided](https://github.com/oasis-open/cti-taxii-server/pull/99#issuecomment-588288204) that if callers want to set the encoding themselves, they should create a file-like object and not pass a string filename.  That way, they can set the file-like object up however they want, including choosing an encoding.  I [agreed](https://github.com/oasis-open/cti-taxii-server/pull/99#issuecomment-588498330); it made sense to me.  So we decided not to have an `encoding` parameter to the function.

**Pull request merged**